### PR TITLE
Lessen memory usage for wrapping vtk datasets

### LINKED ISSUE #3178: Reader memory  usage/memory usage when wrapping vtkDataSets
As pointed out in https://github.com/pyvista/pyvista-support/issues/500#issuecomment-921108703 and https://github.com/pyvista/pyvista-support/issues/500#issuecomment-921106561, readers seem to duplicate memory usage.

Using this code shows that the mesh data is copied from the reader to the mesh object.  This is only released when deleting the reader object.

```py
import pyvista as pv
from pyvista import examples

from memory_profiler import profile

@profile
def run():
    filename = examples.download_parched_canal_4k(load=False)
    reader = pv.get_reader(filename)
    mesh = reader.read()
    del reader

if __name__ == "__main__":
    run()
```

```txt
$python test.py 
Filename: test.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     6    110.2 MiB    110.2 MiB           1   @profile
     7                                         def run():
     8    110.2 MiB      0.0 MiB           1       filename = examples.download_parched_canal_4k(load=False)
     9    110.4 MiB      0.2 MiB           1       reader = pv.get_reader(filename)
    10    302.8 MiB    192.3 MiB           1       mesh = reader.read()
    11    206.8 MiB    -96.0 MiB           1       del reader
```

### PR DESCRIPTION
**Pull request by MatthewFlamm**
### Overview

 

This PR adds a `deep` kwarg to all pyvista `DataSets` with default `False`.  This keyword is used when passing in a `vtkDataSet`. Some datasets already did this.  Some only did deep copy.

Closes #3178 

### Details

We could also consider using this kwarg when passing in arrays or other, but I haven't touched that logic here.

TODO:
- [x] add more tests

### Breaking Change

Previously, some pyvista `DataSet`s copied vtk `DataSet`s when wrapping them.  This PR changes the default to not copy for all `DataSet`s.  Any changes to the pyvista object will also change the original vtk object.  A kwarg `deep` was added to obtain the prior behavior, if needed.

### CODE REVIEW

**Comment by MatthewFlamm** [comment]
I also tested this PR with the Reader example in #3178, and the memory usage there is fixed.

File: pyvista/core/grid.py [hunk 1]
```diff
@@ -419,6 +426,10 @@ class UniformGrid(_vtk.vtkImageData, Grid, UniformGridFilters):
     origin : iterable, optional
         Origin of the uniform grid.  Defaults to ``(0.0, 0.0, 0.0)``.
 
+    deep : bool, optional
+        Whether to deep copy a vtkImage3Data object.
```
Comment (MatthewFlamm) [inline pyvista/core/grid.py:430]:
```suggestion
        Whether to deep copy a vtkImageData object.
```

File: tests/test_grid.py [hunk 2]
```diff
@@ -1400,3 +1409,21 @@ def test_ExplicitStructuredGrid_compute_connections():
     grid.compute_connections(inplace=True)
     assert 'number_of_connections' in grid.cell_data
     assert np.array_equal(grid.cell_data['number_of_connections'], connections)
+
+
+def test_copy_no_copy_wrap_object(datasets):
+    for dataset in datasets:
+        # different dataset tyoes have different copy behavior for points
```
Comment (MatthewFlamm) [inline tests/test_grid.py:1416]:
```suggestion
        # different dataset types have different copy behavior for points
```

File: tests/test_grid.py [hunk 3]
```diff
@@ -1400,3 +1409,21 @@ def test_ExplicitStructuredGrid_compute_connections():
...
+        # use point data which is common
+        dataset["data"] = np.ones(dataset.n_points)
+        new_dataset = type(dataset)(dataset)
+        new_dataset["data"] += 1
+        assert np.array_equal(new_dataset["data"], dataset["data"])
+
+    for dataset in datasets:
+        # different dataset tyoes have different copy behavior for points
```
Comment (MatthewFlamm) [inline tests/test_grid.py:1424]:
```suggestion
        # different dataset types have different copy behavior for points
```

File: pyvista/core/pointset.py [hunk 4]
```diff
@@ -1297,6 +1297,14 @@ class UnstructuredGrid(_vtk.vtkUnstructuredGrid, PointGrid, UnstructuredGridFilt
     - From cell, offset, and node arrays
     - From a file
 
+    Parameters
+    ----------
+    args : various
```
Comment (MatthewFlamm) [inline pyvista/core/pointset.py:1302]: I don't like this, but I'm not even sure how the inputs for some of these work.  I wanted to put in a placeholder for further improvement.

File: pyvista/core/grid.py [hunk 5]
```diff
@@ -95,6 +95,10 @@ class RectilinearGrid(_vtk.vtkRectilinearGrid, Grid, RectilinearGridFilters):
         ``False``. If ``True``, an error is raised if there are any duplicate
         values in any of the array-valued input arguments.
 
+    deep : bool, optional
+        Whether to deep copy a vtkRectilinearGrid object.
```
Comment (akaszynski) [inline pyvista/core/grid.py:99]:
```suggestion
        Whether to deep copy a ``vtk.vtkRectilinearGrid`` object.
```

File: pyvista/core/grid.py [hunk 6]
```diff
@@ -419,6 +426,10 @@ class UniformGrid(_vtk.vtkImageData, Grid, UniformGridFilters):
     origin : iterable, optional
         Origin of the uniform grid.  Defaults to ``(0.0, 0.0, 0.0)``.
 
+    deep : bool, optional
+        Whether to deep copy a vtkImageData object.
```
Comment (akaszynski) [inline pyvista/core/grid.py:430]:
```suggestion
        Whether to deep copy a ``vtk.vtkImageData`` object.
```

File: pyvista/core/pointset.py (hunk 4 above)
Comment (akaszynski) [reply pyvista/core/pointset.py:1302]: Out of scope for this PR, but we can add this later. Better than nothing.

We just need to update this in the same way we did with `UniformGrid`.

File: pyvista/core/pointset.py [hunk 7]
```diff
@@ -2224,6 +2241,14 @@ class ExplicitStructuredGrid(_vtk.vtkExplicitStructuredGrid, PointGrid):
     - From a VTU or VTK file
     - From ``dims`` and ``corners`` arrays
 
+    Parameters
+    ----------
+    args : various
+        See below examples.
+    deep : optional
+        Whether to deep copy a vtkUnstructuredGrid object.
```
Comment (akaszynski) [inline pyvista/core/pointset.py:2249]:
```suggestion
        Whether to deep copy a ``vtk.vtkUnstructuredGrid`` object.
```

**Review by akaszynski — APPROVED** [review]
Made a few nitpicks, otherwise LGTM.

**Comment by adam-grant-hendry** [comment]
@MatthewFlamm @akaszynski I love this and thanks for doing it!

Quick question: does a shallow copy just return a reference to the object? If so, will users possibly mutate their data with this change when they're not expecting it? Is this a potentially a breaking change for existing code bases that previously relied on certain data sets performing deep copies by default?

**Comment by MatthewFlamm** [comment]
> so, will users possibly mutate their data with this change when they're not expecting it? Is this a potentially breaking change for existing code bases that previously relied on certain data sets performing deep copies by default?

Yes you have it right that any change to the pyvista object will change the original vtk object.  I'm arguing that this is the expected contract for pyvista use with `deep=False`.  This was implemented for a few DataSets, e.g. `PolyData` and `UnstructuredGrid` but not others.  This PR aligns them all.  You are right that is a user was relying on the old behavior, then it will be a breaking change.  I'll tag it and put in the PR description, thanks for pointing this out!

**Comment by adam-grant-hendry** [comment]
> > so, will users possibly mutate their data with this change when they're not expecting it? Is this a potentially breaking change for existing code bases that previously relied on certain data sets performing deep copies by default?
> 
> Yes you have it right that any change to the pyvista object will change the original vtk object. I'm arguing that this is the expected contract for pyvista use with `deep=False`. This was implemented for a few DataSets, e.g. `PolyData` and `UnstructuredGrid` but not others. This PR aligns them all. You are right that is a user was relying on the old behavior, then it will be a breaking change. I'll tag it and put in the PR description, thanks for pointing this out!

No worries! Thanks for aligning things: I agree we should be using `shallow_copy`. Thank you!

**Pull request merged**